File: frontend/src/api/backups.ts
import { apiDelete, apiGet, apiPost } from "./client";

export type BackupRecord = {
  backup_id: number;
  project_id: number;
  backup_type: string;
  file_name: string;
  file_path: string;
  file_size: number;
  status: string;
  created_at: string;
  error_code: string | null;
  error_message: string | null;
  download_url: string;
};

export type ProjectBackupResult = {
  backup_id: number;
  project_id: number;
  file_name: string;
  file_path: string;
  file_size: number;
  created_at: string;
  download_url: string;
};

export type RestoreRecord = {
  restore_id: number;
  source_backup_id: number | null;
  source_project_name: string | null;
  new_project_id: number | null;
  restore_mode: string;
  status: string;
  created_at: string;
  error_code: string | null;
  error_message: string | null;
};

export type RestoreBackupResult = {
  restore_id: number;
  backup_id: number;
  source_project_name: string;
  new_project_id: number;
  new_project_name: string;
  status: string;
  restored_counts: Record<string, number>;
  created_at: string;
};

export type BackupVerifyResult = {
  backup_id: number | null;
  valid: boolean;
  warnings: string[];
  errors: string[];
  counts: {
    manifest_files?: number;
    missing_files?: number;
    checksum_failed?: number;
    manifest_count_mismatches?: number;
  };
  summary: {
    has_manifest: boolean;
    has_project_data: boolean;
    file_count: number;
    missing_file_count: number;
    checksum_failed_count: number;
  };
};

export type ProjectIntegrityResult = {
  project_id: number;
  valid: boolean;
  warnings: string[];
  errors: string[];
  path_check: {
    invalid_paths: number;
    missing_files: number;
  };
  counts: Record<string, number>;
};

export function createProjectBackup(projectId: number): Promise<ProjectBackupResult> {
  return apiPost<ProjectBackupResult>(`/api/projects/${projectId}/backup`, undefined, {
    timeoutMs: 120_000
  });
}

export function listBackups(): Promise<BackupRecord[]> {
  return apiGet<BackupRecord[]>("/api/backups");
}

export function listProjectBackups(projectId: number): Promise<BackupRecord[]> {
  return apiGet<BackupRecord[]>(`/api/projects/${projectId}/backups`);
}

export function downloadBackupUrl(backupId: number) {
  return `/api/backups/${backupId}/download`;
}

export function downloadBackup(backupId: number) {
  window.location.href = downloadBackupUrl(backupId);
}

export function deleteBackup(backupId: number): Promise<void> {
  return apiDelete<void>(`/api/backups/${backupId}`);
}

export function restoreBackupAsNewProject(backupId: number): Promise<RestoreBackupResult> {
  return apiPost<RestoreBackupResult>(
    `/api/backups/${backupId}/restore`,
    { restore_mode: "new_project" },
    { timeoutMs: 120_000 }
  );
}

export function listRestores(): Promise<RestoreRecord[]> {
  return apiGet<RestoreRecord[]>("/api/restores");
}

export function verifyBackup(backupId: number): Promise<BackupVerifyResult> {
  return apiGet<BackupVerifyResult>(`/api/backups/${backupId}/verify`);
}

export function checkProjectIntegrity(projectId: number): Promise<ProjectIntegrityResult> {
  return apiGet<ProjectIntegrityResult>(`/api/projects/${projectId}/integrity-check`);
}

File: backend/api/backups.py
from fastapi import APIRouter, Depends, Response, status
from fastapi.responses import FileResponse
from sqlalchemy.orm import Session

from backend.core.database import get_db
from backend.schemas.backup import (
    BackupRecordRead,
    BackupVerifyResult,
    ProjectIntegrityResult,
    ProjectBackupResult,
    RestoreBackupRequest,
    RestoreBackupResult,
    RestoreRecordRead,
)
from backend.services import backup_integrity_service, backup_service, restore_service

router = APIRouter(prefix="/api", tags=["backups"])


@router.post("/projects/{project_id}/backup", response_model=ProjectBackupResult)
def create_project_backup(project_id: int, db: Session = Depends(get_db)) -> ProjectBackupResult:
    return backup_service.create_project_backup(db, project_id)


@router.get("/backups", response_model=list[BackupRecordRead])
def list_backups(db: Session = Depends(get_db)) -> list[BackupRecordRead]:
    return backup_service.list_backups(db)


@router.get("/projects/{project_id}/backups", response_model=list[BackupRecordRead])
def list_project_backups(project_id: int, db: Session = Depends(get_db)) -> list[BackupRecordRead]:
    return backup_service.list_project_backups(db, project_id)


@router.get("/backups/{backup_id}/download")
def download_backup(backup_id: int, db: Session = Depends(get_db)) -> FileResponse:
    path, file_name = backup_service.backup_file_path(db, backup_id)
    return FileResponse(path, filename=file_name, media_type="application/zip")


@router.get("/backups/{backup_id}/verify", response_model=BackupVerifyResult)
def verify_backup(backup_id: int, db: Session = Depends(get_db)) -> BackupVerifyResult:
    return backup_integrity_service.verify_backup_record(db, backup_id)


@router.get("/projects/{project_id}/integrity-check", response_model=ProjectIntegrityResult)
def check_project_integrity(project_id: int, db: Session = Depends(get_db)) -> ProjectIntegrityResult:
    return backup_integrity_service.check_project_integrity(db, project_id)


@router.delete("/backups/{backup_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_backup(backup_id: int, db: Session = Depends(get_db)) -> Response:
    backup_service.delete_backup(db, backup_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/backups/{backup_id}/restore", response_model=RestoreBackupResult)
def restore_backup(
    backup_id: int,
    payload: RestoreBackupRequest,
    db: Session = Depends(get_db),
) -> RestoreBackupResult:
    return restore_service.restore_project_backup_as_new_project(db, backup_id, payload.restore_mode)


@router.get("/restores", response_model=list[RestoreRecordRead])
def list_restores(db: Session = Depends(get_db)) -> list[RestoreRecordRead]:
    return restore_service.list_restores(db)
